### DISCUSSION TITLE
How to set one target to see modules from another?

### DISCUSSION THREAD

**Discussion by kridenberg** [xmake-io/xmake#5229] [Q&A] [answered]
Sorry to bother the community again, but I have minimal examples that are compiled under CMake but do not compile under XMake with CLang. A minimal example builds two targets. Both targets are static libraries, and both export their primary modules. The target2 target imports the primary module from target1, but something is going wrong, and target2 does not see modules from target1.

[xmake.zip](https://github.com/user-attachments/files/15896703/xmake.zip)

1. Testing was done with and without the #5224 patch.
2. Testing was done with the LLVM trunk.
3. It is not very obvious from the documentation how to set up this.

UPD: I found that test https://github.com/xmake-io/xmake/tree/master/tests/projects/c%2B%2B/modules/user_headerunit2

But it also does not compile. With header units, it fails with header units that are not known, and without header units, with error I receive on my test.

**Comment by kridenberg** [comment]
Additionally, the stl_headerunit test from https://github.com/xmake-io/xmake/tree/master/tests/projects/c%2B%2B/modules/stl_headerunit cannot be compiled with clang trunk with the error: "src\hello.mpp:5:8: error: header file <iostream> (aka '<absolute path to the header>') cannot be imported because it is not known to be a header unit"

**Reply by waruqi** [reply]
> Additionally, the stl_headerunit test from https://github.com/xmake-io/xmake/tree/master/tests/projects/c%2B%2B/modules/stl_headerunit cannot be compiled with clang trunk with the error: "src\hello.mpp:5:8: error: header file (aka '') cannot be imported because it is not known to be a header unit"

clang-scan-deps does not yet support headerunit very well now. We need to wait for its support.

Or try https://github.com/xmake-io/xmake/blob/872ddbd0ce9e645a079e1bef02dc2d3424ad4879/tests/projects/c%2B%2B/modules/test_headerunits.lua#L57

`build.c++.clang.fallbackscanner` policy will use xmake deps scanner instead of `clang-scan-deps`.

**Reply by kridenberg** [reply]
The fallback scanner helped. Thanks.

**Comment by waruqi** [comment]
> The target2 target imports the primary module from target1, but something is going wrong, and target2 does not see modules from target1.

I haven't had time to debug it yet.

**Reply by waruqi** [reply]
@Arthapz Can you help me to see this issue?

**Reply by waruqi** [reply, marked as answer]
@kridenberg You should set `{public = true}` to export `.ixx` module interfaces for target_1

```lua
target("test_target_1")
    add_files("a/file1.ixx", {public = true})
```

then test_target_2 will get it. see https://github.com/xmake-io/xmake/blob/872ddbd0ce9e645a079e1bef02dc2d3424ad4879/tests/projects/c%2B%2B/modules/staticlib/xmake.lua#L6

**Reply by kridenberg** [reply]
I assume that is the correct answer, thanks.
I experimented with that, and it was not helpful at the beginning. Maybe previously I did something else incorrectly in addition to that. Thanks.

**Reply by kridenberg** [reply]
So, after all those problems were fixed, I managed to compile the project.
Thanks for all your efforts and help.

**Reply by Arthapz** [reply]
and if u make a module only target, u can `set_kind("moduleonly")` which mark all modules public by default
